Implement scrape-time rule evaluation

fpetkovski opened the pull request:
This commit extends Prometheus with the option to run aggregation rules
during scrape time, before relabeling takes place.
All rules are executed independently for each target in the context of
individual scrapes. Newly calculated samples are
committed to TSDB atomically with all samples from that scrape.

Metric relabeling can also be applied on series recorded from scrape rules.

Signed-off-by: fpetkovski <filip.petkovsky@gmail.com>

cc @bwplotka @roidelapluie the implementation is based on the discussion from the last dev summit.

<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

metalmatze commented:
Great addition! Looking forward to this.
Seems like some files aren't gofumpt'ed. 
Here's how to automatically run it from your editor: https://github.com/mvdan/gofumpt#installation

fpetkovski commented:
Thanks for the tip Matthias, should be fixed now 👍

prymitive commented on scrape/rules.go:
```diff
@@ -0,0 +1,197 @@
+		}
+
+		for _, s := range samples {
+			lbls := s.Metric.WithoutLabels("__name__")
```
You can use `labels.MetricName` instead of hardcoding `__name__` here.

https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L27

fpetkovski replied:
Good point, updated.

prymitive commented on cmd/prometheus/main.go:
```diff
@@ -191,6 +192,9 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 				level.Info(logger).Log("msg", "Automatically set GOMAXPROCS to match Linux container CPU quota")
+			case "scrape-rules":
+				c.enableScrapeRules = true
+				level.Info(logger).Log("msg", "Experimental scrape rules enabled")
```
It appears that this is only for recording rules, is it worth mentioning it here?

fpetkovski replied:
Thanks, I updated the description to reflect that. We can also change the feature flag name to `scrape-recording-rules` if that makes it more clear what the feature is about.

fpetkovski commented on scrape/rules.go:
```diff
@@ -0,0 +1,197 @@
+	return &ruleEngine{
+		targetLabels: targetLabels,
+		rules:        rules,
+		engine: promql.NewEngine(promql.EngineOpts{
```
Good point. I initially added this here to avoid changing lots of constructors, but forgot to clean it up. Should be updated now.

bwplotka commented on config/config.go:
```diff
@@ -447,8 +449,46 @@ type ScrapeConfig struct {
 	MetricRelabelConfigs []*relabel.Config `yaml:"metric_relabel_configs,omitempty"`
+
+	// List of rules to execute at scrape time.
+	RuleConfigs []*ScrapeRuleConfig `yaml:"rule_configs,omitempty"`
```
What are the semantics here in regards to `MetricRelabelConfig` and `RelabelConfig`. Can we document it maybe here?

Essentially what's the order of those relabellings.

I wonder if decision to put rule configs here will require another relabel config - so we can do relabelling before scrape, after scrape and after recording rules.

Some use cases:
A. I want to drop some series/labels on scrape or rename them, before recording
B. I want to drop some series after recording, so I have only recorded series.

Another alternative is to bring rule configuration into Metric relabelling 🤔 (probably less intuitive and harder to use) or do Rule recording BEFORE MetricRelabelConfig if we don't see clear use case like `A` - it might be arguable.

WDYT? cc @roidelapluie @prymitive

fpetkovski replied:
On second thought, I think we can allow applying metric relabelings to series from scrape rules. The order of operations would then be:
* relabel
* evaluate scrape rules
* metrics relabel

This will allow for use cases where expensive series are aggregated into a record, and then during metric relabeling the original series are dropped and the record is renamed to have the name of the original metric. 

I updated the implementation to allow relabeling recorded samples and documented this behavior in the docs.

bwplotka reviewed (commented):
Thanks for this!

I was initially sceptical seeing this: 
![image](https://user-images.githubusercontent.com/6950331/162734565-83da5be0-ea79-475c-8c1b-4c6aa1d515ea.png)
since it smells like code spaghetti and initially looks like we allow querying whole storage with scrape recording rules, but I was wrong. We only scrape batches coming from a single target - so all good.

Generally, I like it but did not review PR in detail yet. There is one big question around the configuration part of it, commented below.

cc @beorn7 @roidelapluie

beorn7 commented:
Great to see this happening. It's in my review queue, but I don't know when I'll get to it. Feel free to move forward without me looking at it.

sharadgaur commented:
cc @csmarchbanks

juliusv commented on config/config.go:
```diff
@@ -446,8 +448,46 @@ type ScrapeConfig struct {
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		switch node.(type) {
+		case *parser.MatrixSelector:
+			err = errors.New("matrix selectors are not allowed in scrape rule expressions")
```
Historically there was always the idea that some day scrape-time rules could allow at least instantaneous rate (`irate()`) calculations, which would require looking back one more sample at least. You could say that dimensional aggregation in scrape-time rules is much more important than that, but the problem is that a lot of metrics are counters, which can't be aggregated (or used in other useful ways) before being rate-ed.

However, now that I think about it more, counters within a *single* target would almost always reset together at the same time (unless you are exposing state for multiple third parties keeping counter states on their own), so maybe it would allow you to get around the "rate-before-sum" rule in this special case.

brian-brazil replied:
The main problem there is a counter being removed, so rate-before-sum still applies. To solve it properly, you need to auto-add the target labels to the selector and fetch the data from the tsdb.

fpetkovski replied:
Another case where a counter could reset are overflows. Would that also cause problems with sum before rate?

brian-brazil replied:
It would, you can only violate rate-before-sum if you know exactly what you're doing and it thus can never be general advice.

fpetkovski replied:
I also wonder if fetching data from tsdb can be avoided by keeping samples from previous scrapes cached in each scrape loop. If we want to use samples from tsdb, then we can't drop series after aggregation and we still have to pay for having them in tsdb. This kind of defeats the purpose of aggregating at scrape time and I am not sure if it is much better than having a regular recording rule.

brian-brazil replied:
The purpose of scrape-time evaluation is to spread CPU load around, so the amount of data a recording rule can handle is less of a limiting factor. Recording rules are (basically) single threaded, whereas many scrapes can happen concurrently.
That the results will also have the same timestamp as the scrape can also be handy.

prymitive replied:
One common want in my org is to have a per cluster aggregate without the need to store all the per instance metrics. So my initial hope was for that - build an aggregate from scrape results, save it and discard all those metrics. But because scrapes are per target it won't work.
If it was possible then it would be a big win for some services we run since it's just too expensive to scrape all per instance metrics just so that you can build an aggregate.
That being said it's entirely possible that same could be achieved with ability to set per metric retention period since we could keep only enough samples for per instance metrics to allow aggregation rules to run.

beorn7 replied:
More thoughts about counter resets: We don't really get overflows with our float64 counters. Instead, we run into "loss of precision" areas, which could be taken as a reason to reset an individual counter. I haven't heard of anyone doing this yet, but it could be done. And, fun fact, the upcoming new histograms use a similar trick: Resetting the histogram to shed entropy once too many buckets have been populated.

I would suggest to keep matrix selectors out of the game for now and think about them in a later iteration of the feature.

beorn7 replied:
Right, that's where the counter comes from a different monitoring system.

Anyway, bottom line is: counter resets can happen for a whole lot of reasons, not only if the binary restarts. It's really hard to rule them out.

juliusv commented on config/config.go:
```diff
@@ -446,8 +448,46 @@ type ScrapeConfig struct {
 	MetricRelabelConfigs []*relabel.Config `yaml:"metric_relabel_configs,omitempty"`
+
+	// List of rules to execute at scrape time.
+	RuleConfigs []*ScrapeRuleConfig `yaml:"scrape_rule_configs,omitempty"`
```
Not sure what to do about it, but it will a bit confusing that normal rules are configured using separate files, while scrape-time rules are configured inline. Probably that's fine for now though.

juliusv reviewed (commented):
Just some comments witthout having done a full review yet.

Nexucis commented on web/ui/module/codemirror-promql/src/grammar/parser.js:
```diff
@@ -0,0 +1,19 @@
+// This file was generated by lezer-generator. You probably shouldn't edit it.
```
can you please remove this file.

It's not supposed to be versioned.

fpetkovski replied:
Thanks, good catch.

Nexucis commented on web/ui/module/codemirror-promql/src/grammar/parser.terms.js:
```diff
@@ -0,0 +1,148 @@
+// This file was generated by lezer-generator. You probably shouldn't edit it.
```
same for this one.
Thanks!

amuraru commented on docs/configuration/configuration.md:
```diff
@@ -2714,6 +2719,28 @@ anchored on both ends. To un-anchor the regex, use `.*<regex>.*`.
+
+Metric relabelings can be applied to series generated by scrape rules.
+
+A potential use of scrape rules is to aggregate away expensive series before they are ingested in Prometheus.
```
the source series would need to be removed in this case. Is it worth adding this note here?

fpetkovski replied:
Good idea, added a note.

GiedriusS commented:
Ping :hugs: this came up again in https://cloud-native.slack.com/archives/C167KFM6C/p1670016589800069

beorn7 commented:
@fpetkovski many thanks for picking this up again. Do you still have questions about how to proceed, or is it time for an in-depth review? (@bwplotka @roidelapluie @juliusv is one of you up for it? I'm already very busy with my growing review queue of native histogram related PRs…)

fpetkovski commented:
I think most high level comments should be addressed now and an in-depth review would be great. @roidelapluie had some concerns about scalability of the approach in large environments and I can try to do some benchmarking next week.

Also worth noting that native histograms are not yet supported. I can either add them to this PR or make a follow up.

roidelapluie commented on cmd/prometheus/main.go:
```diff
@@ -611,11 +615,26 @@ func main() {
 
+	opts := promql.EngineOpts{
+		Logger:                   log.With(logger, "component", "query engine"),
+		Reg:                      prometheus.DefaultRegisterer,
```
Do you think this would work out? it feels like we would need so somehow wrap the metrics.

bboreham commented:
@SuperQ I believe you had ideas in this area - does this conflict?
@bwplotka you said in bug-scrub you would take a look.

bwplotka commented:
@SuperQ the documentation is in this PR: https://github.com/prometheus/prometheus/blob/bd4a6a7e8d1cef2f99a9f22cdce77bb1d0d50339/docs/configuration/configuration.md#scrape_rule_configs, but true, it might be amazing to show end user example of this in the PR description.

bboreham commented:
If I understood correctly, this proposal could work to aggregate gauges but not counters, since there is no mechanism to handle resets. 

Either way, the situation should be made clear.

fpetkovski commented:
> If I understood correctly, this proposal could work to aggregate gauges but not counters, since there is no mechanism to handle resets.

This was discussed briefly here https://github.com/prometheus/prometheus/pull/10529#discussion_r878827957. In theory if you're certain that all of your counters will reset at the same time, which is common when they're on the same target, then you could aggregate them before applying a rate. For example, it should be safe to aggregate CPU usage for all cores from the node exporter.

bboreham commented:
OK the other thing I missed is that you mean for these rules to use as input only the data received in one scrape. That seems to rule out most cases where people want to aggregate.

The only case I can think of where I would have wanted this is per-core CPU usage.

Can you edit the description of this PR so it mentions things like this please?

beorn7 commented:
While in practice most counters on the same target reset at the same time, that's not generally true. Following cases come to mind:
- Native histograms reset on purpose now and then to shed entropy.
- Counter overflows might still happen (not really with float64, and unlikely with int64, but some counters are still backed by int32, where this can happen in practice).
- A target might remove a counter completely (deregister it) and bring it back shortly after, starting at zero again.

bboreham commented:
Hello from the bug-scrub!  @fpetkovski do you think you will return to this?
I note that the related project Grafana Alloy is looking at a similar feature - https://github.com/grafana/alloy/issues/414.

fpetkovski commented:
@bboreham yes we still want this and I can come back to it. I'll try to take a look this week.

fpetkovski committed:
Implement rule evaluations

This commit extends Prometheus with the option to run aggregation rules
during scrape time, before relabeling takes place.

All rules are executed independently for each target in the context of
individual scrapes. Newly calculated samples are
committed to TSDB atomically with all samples from that scrape.

Metric relabeling can also be applied on series recorded from scrape rules.

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Run lint

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Fix dedupelabels

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Further restrict query expressions

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Update stale markers after recording rules

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Fix lint

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski commented:
@roidelapluie I've addressed all your comments so far, PTAL again!

fpetkovski committed:
Rename and validate scrape_rules

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

fpetkovski committed:
Capture non-relabeled metrics from scrape

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

roidelapluie commented:
WDYT of https://github.com/prometheus/proposals/pull/67 ?

bboreham commented:
Hello from the bug-scrub!

Looks like this is now waiting on a resolution of https://github.com/prometheus/proposals/pull/67
